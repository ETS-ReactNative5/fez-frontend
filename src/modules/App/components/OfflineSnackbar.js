import React, {PureComponent} from 'react';
import {PropTypes} from 'prop-types';
import Snackbar from 'material-ui/Snackbar';
import ActionCheckCircle from 'material-ui/svg-icons/action/check-circle';
import AlertError from 'material-ui/svg-icons/alert/error';

export default class OfflineSnackbar extends PureComponent {
    static propTypes = {
        locale: PropTypes.object.isRequired
    };

    static defaultProps = {
        locale: {
            open: false,
            message: '',
            autoHideDuration: null
        }
    };

    constructor(props) {
        super(props);
        this.state = {
            online: typeof navigator.onLine === 'boolean' ? navigator.onLine : true,
            hasDisconnected: false
        };
    }

    componentDidMount() {
        window.addEventListener('online', () => {
            this.setState({online: true});
        });
        window.addEventListener('offline', () => {
            this.setState({online: false, hasDisconnected: true});
        });
    }

    renderMessage = (message, icon) => {
        return (
            <div className={`columns is-gapless connectionStatus is-mobile ${this.state.online ? 'online' : 'offline'}`}>
                <div className="column"/>
                <div className="column is-narrow">{icon}</div>
                <div className="column is-narrow">{message}</div>
                <div className="column"/>
            </div>
        );
    };

    render() {
        const locale = this.props.locale;
        const snackbarProps = this.state.online ?
            {...locale.online, message: this.renderMessage(locale.online.message, <ActionCheckCircle/>)} :
            {...locale.offline, message: this.renderMessage(locale.offline.message, <AlertError/>)};
        return  (
            <div className="offlineSnackbar">
                {
                    (!this.state.online || (this.state.online && this.state.hasDisconnected)) &&
                    <Snackbar {...snackbarProps} />
                }
            </div>
        );
    }
}
