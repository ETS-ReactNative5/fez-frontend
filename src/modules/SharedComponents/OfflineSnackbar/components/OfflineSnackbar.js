import React, {PureComponent} from 'react';
import Snackbar from 'material-ui/Snackbar';
import ActionCheckCircle from 'material-ui/svg-icons/action/check-circle';
import AlertError from 'material-ui/svg-icons/alert/error';
import {locale} from 'locale';

export default class OfflineSnackbar extends PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            open: !navigator.onLine,
            online: navigator.onLine
        };
    }

    componentDidMount() {
        window.addEventListener('online', this.updateOnlineState);
        window.addEventListener('offline', this.updateOnlineState);
    }

    componentWillUnmount() {
        window.removeEventListener('online', this.updateOnlineState);
        window.removeEventListener('offline', this.updateOnlineState);
    }

    updateOnlineState = () => {
        this.setState({open: true, online: navigator.onLine});
    };

    renderMessage = (message, icon) => {
        return (
            <div className={`columns is-gapless connectionStatus is-mobile ${this.state.online ? 'online' : 'offline'}`}>
                <div className="column"/>
                <div className="column is-narrow">{icon}</div>
                <div className="column is-narrow">{message}</div>
                <div className="column"/>
            </div>
        );
    };

    handleRequestClose = (reason) => {
        // MUI hack to prevent the snackbar from being hidden by clicking/touchTapping away
        if (reason !== 'clickaway') {
            this.setState({open: false});
        }
    };

    render() {
        const txt = locale.global.offlineSnackbar;
        const snackbarProps = this.state.online ?
            {...txt.online, message: this.renderMessage(txt.online.message, <ActionCheckCircle/>)} :
            {...txt.offline, message: this.renderMessage(txt.offline.message, <AlertError/>)};
        return  (
            <div className="offlineSnackbar">
                <Snackbar
                    {...snackbarProps}
                    open={this.state.open}
                    onRequestClose={this.handleRequestClose}
                    onActionTouchTap={this.handleRequestClose}
                />
            </div>
        );
    }
}
