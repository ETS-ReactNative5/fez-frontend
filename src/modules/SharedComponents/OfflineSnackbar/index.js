export {default as OfflineSnackbar} from './components/OfflineSnackbar';
