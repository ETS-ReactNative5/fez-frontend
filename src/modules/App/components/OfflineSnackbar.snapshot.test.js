jest.dontMock('../components/OfflineSnackbar');

import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import OfflineSnackbar from '../components/OfflineSnackbar';

function setup({...props}) {
    return shallow(<OfflineSnackbar {...props} />);
}

describe('OfflineSnackbar snapshots test', () => {

    const locale = {
        online: {
            open: true,
            message: 'Online',
            autoHideDuration: 5000
        },
        offline: {
            open: true,
            message: 'Offline',
        }
    };

    it('renders empty component', () => {
        const wrapper = setup({locale});
        wrapper.instance().setState({online: true, hasDisconnected: false});
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders offline snackbar', () => {
        const wrapper = setup({locale});
        wrapper.instance().setState({online: false, hasDisconnected: false});
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders back online snackbar', () => {
        const wrapper = setup({locale});
        wrapper.instance().setState({online: true, hasDisconnected: true});
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
