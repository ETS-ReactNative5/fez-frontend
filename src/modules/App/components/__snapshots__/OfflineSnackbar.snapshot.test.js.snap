// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfflineSnackbar snapshots test renders back online snackbar 1`] = `
<div
  className="offlineSnackbar"
>
  <Snackbar
    autoHideDuration={5000}
    message={
      <div
        className="columns is-gapless connectionStatus is-mobile online"
      >
        <div
          className="column"
        />
        <div
          className="column is-narrow"
        >
          <ActionCheckCircle />
        </div>
        <div
          className="column is-narrow"
        >
          Online
        </div>
        <div
          className="column"
        />
      </div>
    }
    open={true}
  />
</div>
`;

exports[`OfflineSnackbar snapshots test renders nothing 1`] = `
<div
  className="offlineSnackbar"
/>
`;

exports[`OfflineSnackbar snapshots test renders offline snackbar 1`] = `
<div
  className="offlineSnackbar"
>
  <Snackbar
    message={
      <div
        className="columns is-gapless connectionStatus is-mobile offline"
      >
        <div
          className="column"
        />
        <div
          className="column is-narrow"
        >
          <AlertError />
        </div>
        <div
          className="column is-narrow"
        >
          Offline
        </div>
        <div
          className="column"
        />
      </div>
    }
    open={true}
  />
</div>
`;
